src/main/java/org/spongepowered/api/service/permission/Subject.java

```diff
@@ -138,34 +150,55 @@
      */
     SubjectData getTransientSubjectData();
 
+    /**
+     * Test whether the subject is permitted to perform an action given as the
+     * given permission string.
+     *
+     * <p>This must return the same value as {@link #hasPermission(String, Cause)}
+     * called with the phase tracker's current cause.
+     *
+     * @param permission The permission string
+     * @return True if permission is granted
+     */
+    default boolean hasPermission(final String permission) {
+        return this.hasPermission(permission, Sponge.getServer().getCauseStackManager().getCurrentCause());
```

Comment:
I have some concerns about the move to using `Cause` instead of `Set<Context` for Subject query methods.

Loose example:

```java
Subject s1 = service.getUserSubjects().loadSubject("--some-uuid--").join();
Subject s2 = service.getUserSubjects().loadSubject("--some-other-uuid--").join();

boolean c1 = s1.hasPermission("some.permission");
boolean c2 = s2.hasPermission("some.permission");
```

It seems reasonable to expect that the `hasPermission` method calls will use cause/context corresponding to the Subject (and associated object) being queried, however, unless the user pushes these onto the Cause stack beforehand, that won't actually happen.

Maybe the fix is for these methods to push `this` and `this.getAssociatedObject()` onto the stack before getting the current cause stack?

Also, the methods can be called async, so we need to account for that too.